motion_computation/src/motion_computation_node.cpp

```diff
@@ -13,52 +13,78 @@
 
   void MotionComputationNode::initialize()
   {
     // Load parameters
     double step = 0.1;
+    double mobility_step = 0.1;
     double period = 2.0;
     double ax = 9.0;
     double ay = 9.0;
     double process_noise_max = 1000.0;
     double drop_rate = 0.9;
+    int external_object_prediction_mode = 1; //sensor only mode
 
     pnh_.param<double>("prediction_time_step", step, step);
+    pnh_.param<double>("mobility_path_time_step", mobility_step, mobility_step);
     pnh_.param<double>("prediction_period", period, period);
     pnh_.param<double>("cv_x_accel_noise", ax, ax);
     pnh_.param<double>("cv_y_accel_noise", ay, ay);
     pnh_.param<double>("prediction_process_noise_max", process_noise_max, process_noise_max);
     pnh_.param<double>("prediction_confidence_drop_rate", drop_rate, drop_rate);
+    pnh_.param<int>("external_object_prediction_mode", external_object_prediction_mode, external_object_prediction_mode);
 
     motion_worker_.setPredictionTimeStep(step);
+    motion_worker_.setMobilityPathPredictionTimeStep(mobility_step);
     motion_worker_.setPredictionPeriod(period);
     motion_worker_.setXAccelerationNoise(ax);
     motion_worker_.setYAccelerationNoise(ay);
     motion_worker_.setProcessNoiseMax(process_noise_max);
     motion_worker_.setConfidenceDropRate(drop_rate);
-
+    motion_worker_.setExternalObjectPredictionMode(external_object_prediction_mode);
+    motion_worker_.setECEFToMapTransform(lookupECEFtoMapTransform());
+    
     // Setup pub/sub
     motion_comp_sub_=nh_.subscribe("external_objects",1,&MotionComputationWorker::predictionLogic,&motion_worker_);
     carma_obj_pub_=nh_.advertise<cav_msgs::ExternalObjectList>("external_object_predictions", 2);
+    mobility_path_sub_=nh_.subscribe("incoming_mobility_path",20,&MotionComputationWorker::mobilityPathCallback,&motion_worker_); // 20 is most number of vehicles in our immeadiate vicinity which might ever need to be tracked.
   }
 
-  void MotionComputationNode::publishObject(const cav_msgs::ExternalObjectList& obj_pred_msg)
+  void MotionComputationNode::publishObject(const cav_msgs::ExternalObjectList& obj_pred_msg) const
   {
-  carma_obj_pub_.publish(obj_pred_msg);
+    carma_obj_pub_.publish(obj_pred_msg);
   }
 
   void MotionComputationNode::run()
   {
     initialize();
-    nh_.setSpinRate(20);
-    nh_.spin();
+    ros::CARMANodeHandle::setSpinRate(20);
+    ros::CARMANodeHandle::spin();
+  }
+
+  tf2::Transform MotionComputationNode::lookupECEFtoMapTransform()
+  {
+    tf2::Transform map_in_earth;
+    try
+    {
+      std::string errstr;
+      if (tf_buffer_.canTransform("earth", "map", ros::Time(0), ros::Duration(20), &errstr)) //20 second timeout
+        tf2::convert(tf_buffer_.lookupTransform("earth", "map", ros::Time(0)).transform, map_in_earth); //save to local copy of transform
+      else
+        ROS_ERROR_STREAM("Look up transform timed out: " << errstr);
```

Comment:
We need to be throwing an exception here. We cannot allow the vehicle to operate without object detection.

Reply:
More specifically a system alert needs to occur. I believe this could be most cleanly achieved with a call to CARMANodeHandle::handleException since this initialize function is probably not wrapped in the default handler.